### PR TITLE
improved support for TRiP98 BDO tags

### LINKED ISSUE #366: DDD files may get E/amu and particle A,Z from future BDO tags
possibly still missing in SH12A, I will notifiy when they are added

### PR DESCRIPTION
**Pull request by grzanka**
Fixes #366

### CODE REVIEW

**Comment by grzanka** [comment]
@nbassler if you want to test, you can install the converter based on code from this PR via:

```
pip install git+https://github.com/DataMedSci/pymchelper@feature/366_ddd_improved
```

then run some of following commands:

```
convertmc tripddd path_to_some_bdo.bdo
convertmc tripddd path_to_some_bdo.bdo --ngauss 1
convertmc tripddd path_to_some_bdo.bdo --ngauss 1 -vv
```

**Comment by nbassler** [comment]
looks proper, also with amus.
```
bassler@kepler:~/Projects/pymchelper$ PYTHONPATH=. python3 pymchelper/run.py  inspect ~/Projects/TRiPbase/src/SH12A_ddd/wdir/1H/RF0MM/100.00/output.bdo 
Tmax_MeV                : '100.0'
Tmax_MeV/amu            : '99.2235723484'
Tmax_MeV/nucl           : '100.0'
error_type              : 'ErrorEstimate.none'
file_corename           : 'output'
file_counter            : '1'
filedate                : 'Sat, 29 Aug 2020 18:32:26 +0200'
geo_name                : 'MyCyl'
geo_unit_ids            : '[10 10 10]'
geo_units               : 'cm;cm'
geotyp                  : 'CYL'
host                    : 'kepler'
mc_code_build_date      : 'Sat, 29 Aug 2020 18:30:02 +0200'
mc_code_version         : 'v0.8.1-17-gb4de72d'
number_of_primaries     : '50000'
page_count              : '1'
projectile_a            : '1'
projectile_code         : '2'
projectile_z            : '1'
user                    : 'bassler'
x                       : 'MeshAxis(n=800, min_val=0.0, max_val=40.0, name='', unit='cm', binning=<BinningType.linear: 0>)'
y                       : 'MeshAxis(n=1, min_val=0.0, max_val=6.2831853071795862, name='', unit='cm', binning=<BinningType.linear: 0>)'
z                       : 'MeshAxis(n=500, min_val=0.0, max_val=50.0, name='', unit='cm', binning=<BinningType.linear: 0>)'
***************************************************************************
Page 0 / 1
	data_unit               : ''
	dettyp                  : 'DDD'
	dif_units               : ''
	diff_axis1              : 'MeshAxis(n=1, min_val=nan, max_val=nan, name='', unit='', binning=<BinningType.linear: 0>)'
	diff_axis2              : 'MeshAxis(n=1, min_val=nan, max_val=nan, name='', unit='', binning=<BinningType.linear: 0>)'
	estimator               : '<pymchelper.estimator.Estimator object at 0x7f7f9dd99340>'
	name                    : ''
	offset                  : '0.0'
	rescale                 : '1.0'
	unit                    : ''
	unit_ids                : '[-1 -1 -1]'
Data min: 0, max: 3051.06
---------------------------------------------------------------------------
```

**Comment by nbassler** [comment]
but something wrong with units of the FWHMs?

```
bassler@kepler:~/Projects/pymchelper$ more output_1H.H2O.MeV09922.ddd
!filetype    ddd
!fileversion   19980520
!filedate      Sat Aug 29 18:38:27 2020
!projectile    1H
!material      H2O
!composition   H2O
!density 1.000000
!energy 99.223572
#
# Created with pymchelper 1.1.0+rev7
#
#   z[g/cm**2] dE/dz[MeV/(g/cm**2)] FWHM1[g/cm**2] factor FWHM2[g/cm**2]
!ddd
0.05 23.3748 6.70656 0.998583 6.94205
0.15 23.4927 6.71816 0.932236 6.95365
0.25 23.6085 6.73404 0.998296 6.96953
```

Untitled

Radial Dose map:
output_e

should rather be 6 mm maybe

**Comment by nbassler** [comment]
how about a dddfixer script? 
- remove NaNs by interpolation if they are not too frequent
- remove noise when beam has stopped
- check and fix outliers, curves should be rather smooth.
- any fixes will be reported to stdout and a logfile.

**Review by nbassler — CHANGES_REQUESTED** [review]
FWHM units out DDD outputs seem wrong.

**Comment by grzanka** [comment]
Can you share SH12A configuration with me ?

In my case DDD for 100 MeV protons looks like this:
```
!filetype    ddd
!fileversion   19980520
!filedate      Sat Aug 29 20:30:39 2020
!projectile    1H
!material      H2O
!composition   H2O
!density 1.000000
!energy 99.223572
#
# Created with pymchelper 1.1.0+rev7
#
#   z[g/cm**2] dE/dz[MeV/(g/cm**2)] FWHM1[g/cm**2] factor FWHM2[g/cm**2]
!ddd
0.0125 0.0941392 nan nan nan
0.0375 0.0941713 0.0951863 0.992374 2.93238
0.0625 0.0955238 0.0914928 0.993623 0.41264
0.0875 0.0961706 0.0907532 0.996027 0.900484
0.1125 0.0971025 0.0980543 0.909583 47.1945
0.1375 0.0993763 0.0990935 0.991702 3.45375
```

**Comment by grzanka** [comment]
Dose map looks also OK:
image

**Comment by nbassler** [comment]
even after the energy -> dose DDD change in SH12A, it still acts strange.

``` 
$ PYTHONPATH=. python3 pymchelper/run.py  --version
1.1.0+rev7
$ PYTHONPATH=. python3 pymchelper/run.py  tripddd --ngauss 2 ~/Projects/TRiPbase/src/SH12A_ddd/wdir/1H/RF0MM/100.00/output.bdo
$ head -16 output_1H.H2O.MeV09922.ddd
!filetype    ddd
!fileversion   19980520
!filedate      Sun Aug 30 00:18:55 2020
!projectile    1H
!material      H2O
!composition   H2O
!density 1.000000
!energy 99.223572
#
# Created with pymchelper 1.1.0+rev7
#
#   z[g/cm**2] dE/dz[MeV/(g/cm**2)] FWHM1[g/cm**2] factor FWHM2[g/cm**2]
!ddd
0.05 317.623 5.06332 0.999318 5.29881
0.15 321.009 5.07304 0.99938 5.30853
0.25 323.965 5.07361 0.999067 5.3091
``` 
FWHMs are way too large.


```
$ PYTHONPATH=. python3 pymchelper/run.py  inspect ~/Projects/TRiPbase/src/SH12A_ddd/wdir/1H/RF0MM/100.00/output.bdo 
Tmax_MeV                : '100.0'
Tmax_MeV/amu            : '99.2235723484'
Tmax_MeV/nucl           : '100.0'
error_type              : 'ErrorEstimate.none'
file_corename           : 'output'
file_counter            : '1'
filedate                : 'Sun, 30 Aug 2020 00:18:35 +0200'
geo_name                : 'MyCyl'
geo_unit_ids            : '[10 10 10]'
geo_units               : 'cm;cm'
geotyp                  : 'CYL'
host                    : 'kepler'
mc_code_build_date      : 'Sun, 30 Aug 2020 00:10:41 +0200'
mc_code_version         : 'v0.8.1-20-g2e670ce'
number_of_primaries     : '200000'
page_count              : '1'
projectile_a            : '1'
projectile_code         : '2'
projectile_z            : '1'
user                    : 'bassler'
x                       : 'MeshAxis(n=800, min_val=0.0, max_val=40.0, name='', unit='cm', binning=<BinningType.linear: 0>)'
y                       : 'MeshAxis(n=1, min_val=0.0, max_val=6.2831853071795862, name='', unit='cm', binning=<BinningType.linear: 0>)'
z                       : 'MeshAxis(n=500, min_val=0.0, max_val=50.0, name='', unit='cm', binning=<BinningType.linear: 0>)'
***************************************************************************
Page 0 / 1
	data_unit               : 'MeV/g'
	dettyp                  : 'DDD'
	dif_units               : 'MeV/g'
	diff_axis1              : 'MeshAxis(n=1, min_val=nan, max_val=nan, name='', unit='', binning=<BinningType.linear: 0>)'
	diff_axis2              : 'MeshAxis(n=1, min_val=nan, max_val=nan, name='', unit='', binning=<BinningType.linear: 0>)'
	estimator               : '<pymchelper.estimator.Estimator object at 0x7feb60064340>'
	name                    : ''
	offset                  : '0.0'
	rescale                 : '1.0'
	unit                    : ''
	unit_ids                : '[40 40 40]'
Data min: 0, max: 383611
---------------------------------------------------------------------------
```

tarball for reproduction:
[100.00.tar.gz](https://github.com/DataMedSci/pymchelper/files/5146046/100.00.tar.gz)

File: pymchelper/run.py [hunk 1]
```diff
@@ -107,31 +91,37 @@ def main(args=sys.argv[1:]):
...
-    parser_tripcube = subparsers.add_parser(Converters.tripcube.name, help='converts to trip98 data cube')
+    parser_tripcube = subparsers.add_parser(Converters.tripcube.name, help='converts to TRiP98 data cube')
     add_default_options(parser_tripcube)
 
-    parser_tripddd = subparsers.add_parser(Converters.tripddd.name, help='converts to trip98 ddd file')
+    parser_tripddd = subparsers.add_parser(Converters.tripddd.name, help='converts to TRiP98 DDD file')
     add_default_options(parser_tripddd)
     parser_tripddd.add_argument("--energy",
-                                help='energy of the beam [MeV/amu] (0 to guess from data)',
+                                help='energy of the beam [MeV/amu] (guess from data if option missing)',
```
Comment (nbassler) [inline pymchelper/run.py:100]: not needed

**Review by nbassler — APPROVED** [review]
ok, but maybe update the --help regarding --energy with something clearer. It is not guessed when shipped with bdo. :)

**Pull request merged**